gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAChallengeEndpoint.java

```diff
@@ -247,7 +258,29 @@ private void verifyCode(RoutingContext routingContext) {
 
         final FactorContext factorCtx = new FactorContext(applicationContext, factorData);
 
-        verify(factorProvider, factorCtx, h -> {
+        verifyAttemptService.checkVerifyAttempt(endUser.getId(), factorId, client, domain).subscribe(verifyAttempt -> {
+                    if (verifyAttempt.isAllowRequest()) {
+                        verify(factorProvider, factorCtx, Optional.of(verifyAttempt), routingContext,
+                                verifyHandler(routingContext, client, endUser, factor, code, factorId, factorProvider, enrolledFactor));
+                    } else {
+                        if (verifyAttemptService.shouldSendEmail(client, domain)) {
+                            new Thread(() -> emailService.send(Template.VERIFY_ATTEMPT, endUser, client)).start();
+                        }
```

Comment:
Don't think we have to create a Thread to send the email. If you don't want to delay the response to the browser, we can do the send after the handleException

Reply:
thread removed